### PR TITLE
Repository cloning (4.2.1)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
Adds a comprehensive ExecPlan for Step 4.2: Clone the requested repository into the sandbox workspace. The plan is a living document guiding the repository-cloning implementation, including constraints, tolerances, risks, progress, decisions, and outcomes.

## Changes
- New file: docs/execplans/4-2-1-repository-cloning.md (Draft, 446 insertions)

## Rationale
This PR introduces a detailed plan to implement repository cloning as a library-owned, reusable primitive. It remains a draft and requires explicit user approval before coding begins, per the plan’s guidance. The ExecPlan specifies:
- cloning format: owner/name into workspace.base_dir with the requested branch checked out
- authentication via GIT_ASKPASS (no credentials in URL or argv)
- input validation requiring explicit repository and a non-empty branch
- observable success criteria and test strategy (unit and behavioural) to be defined in later milestones

## What to review
- Clarity and completeness of constraints, tolerances, risks, and decision log
- Alignment with the project’s design docs and roadmap
- Readability, formatting, and lintability of the document
- Whether the plan reasonably scopes Milestones A–F and their success criteria

## Milestones
- Milestone A: Confirm prerequisites and define the request model
- Milestone B: Add the engine-level repository clone helper
- Milestone C: Add the library wrapper and minimal adapter wiring
- Milestone D: Add unit and behavioural tests
- Milestone E: Update design and user-facing documentation
- Milestone F: Validate and prepare for review

## Risks and constraints
- Prerequisites from Step 3.x (token acquisition and GIT_ASKPASS contract) must exist
- Do not vary workspace.base_dir semantics; explicit clone destination is required
- Avoid exposing tokens or credentials; use GIT_ASKPASS helper
- No new external dependencies
- Changes to public library surface should be minimal and justified

## Next steps
- Await explicit approval to proceed with implementation
- Upon approval, begin Milestone A (prerequisites and request model)

## Documentation impact
- Expects updates to:
  - docs/podbot-design.md (Execution flow and github_clone semantics)
  - docs/podbot-roadmap.md
  - docs/users-guide.md
- Future changes should keep documentation aligned with implemented behavior

## How this will be tested
- This is a documentation-only PR. Concrete tests (unit and behavioural) will be added under Milestones B–D when implementation proceeds.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/f47b24e2-2f73-4d49-b0e3-e010ab8b68e1

## Summary by Sourcery

Documentation:
- Introduce a detailed ExecPlan document for repository cloning, outlining purpose, constraints, tolerances, risks, milestones, and validation steps.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Complex Conditional
src/api/repository_clone.rs: RepositoryRef.parse

### What lead to degradation?
RepositoryRef.parse has 1 complex conditionals with 2 branches, threshold = 2

### Why does this problem occur?
A complex conditional is an expression inside a branch such as an if-statmeent which consists of multiple, logical operations. Example: if (x.started() && y.running()).Complex conditionals make the code even harder to read, and contribute to the Complex Method code smell. Encapsulate them.

### How to fix it?
Apply the [DECOMPOSE CONDITIONAL](https://refactoring.com/catalog/decomposeConditional.html) refactoring so that the complex conditional is encapsulated in a separate function with a good name that captures the business rule. Optionally, for simple expressions, introduce a new variable which holds the result of the complex conditional.

### Helpful refactoring examples
To get a general understanding of what this code health issue looks like - and how it might be addressed - we have prepared some diffs for illustrative purposes.

#### SAMPLE

~~~diff
# complex_conditional.js
 function messageReceived(message, timeReceived) {
-   // Ignore all messages which aren't from known customers:
-   if (!message.sender &&
-       customers.getId(message.name) == null) {
+   // Refactoring #1: encapsulate the business rule in a
+   // function. A clear name replaces the need for the comment:
+   if (!knownCustomer(message)) {
      log('spam received -- ignoring');
      return;
    }
 
-  // Provide an auto-reply when outside business hours:
-  if ((timeReceived.getHours() > 17) ||
-      (timeReceived.getHours() < 8)) {
+  // Refactoring #2: encapsulate the business rule.
+  // Again, note how a clear function name replaces the
+  // need for a code comment:
+  if (outsideBusinessHours(timeReceived)) {
     return autoReplyTo(message);
   }
 
   pingAgentFor(message);
+}
+
+function outsideBusinessHours(timeReceived) {
+  // Refactoring #3: replace magic numbers with
+  // symbols that communicate with the code reader:
+  const closingHour = 17;
+  const openingHour = 8;
+
+  const hours = timeReceived.getHours();
+
+  // Refactoring #4: simple conditional rules can
+  // be further clarified by introducing a variable:
+  const afterClosing = hours > closingHour;
+  const beforeOpening = hours < openingHour;
+
+  // Yeah -- look how clear the business rule is now!
+  return afterClosing || beforeOpening;
 }
~~~

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
tests/bdd_repository_cloning_helpers/assertions.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: the_checked_out_branch_is,the_workspace_path_is

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: src/api/repository_clone.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+fn invalid_repository_ref() -> crate::error::PodbotError {
+    ConfigError::InvalidValue {
+        field: String::from("repo"),
+        reason: String::from("expected repository in owner/name form"),
+    }
+    .into()
+}
+
+fn has_invalid_segments(owner: &str, name: &str) -> bool {
+    owner.is_empty() || name.is_empty() || name.contains('/')
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/repository_clone.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/engine/connection/repository_clone/mod.rs [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+fn github_remote(request: &RepositoryCloneRequest<'_>) -> String {
+    format!(
+        "https://github.com/{}/{}.git",
+        request.repository_owner, request.repository_name
+    )
+}
+
+fn validate_non_empty(field: &str, value: &str) -> Result<(), PodbotError> {
+    if value.trim().is_empty() {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/repository_clone/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 7 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Unit Architecture | ❌ Error | RepositoryCloneRequest accepts raw strings instead of typed validators, duplicating validation logic and obscuring fallibility at the boundary. Engine revalidates what the API already validated. | Use typed fields in RepositoryCloneRequest: &RepositoryRef, &BranchName, &WorkspacePath instead of raw &str. Remove engine validators. This enforces API semantics at query boundary and clarifies command responsibilities. |
| Developer Documentation | ⚠️ Warning | ExecPlan and architectural subsystem well documented, but two review comments unaddressed: grammar fix and Step reference correction. | Fix grammar in docs/developers-guide.md section 2.1 ("This ignore" → "This ignore rule") and Step reference in docs/users-guide.md line 86 ("Step 2.6.4" → "Step 2.6.3"). |
| Testing (Property / Proof) | ⚠️ Warning | WorkspacePath introduces an absolute-path invariant over string inputs but lacks property tests. RepositoryRef and BranchName both have proptest coverage; WorkspacePath should too. | Add proptest property test for WorkspacePath verifying that only absolute paths (starting with /) pass validation and all relative/empty inputs are rejected across arbitrary Unicode input ranges. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | New stable API types lack trybuild compile-time tests; compile\_contract.rs omits repository cloning signatures from stable API coverage. | Create tests/ui/stable\_repository\_clone\_signatures.rs with trybuild compile-pass test for RepositoryRef, BranchName, WorkspacePath, and clone\_repository\_into\_workspace; add to compile\_contract.rs. |
| Domain Architecture | ⚠️ Warning | Domain violation: `RepositoryCloneRequest` takes raw strings instead of typed validators. Validation duplicates across API and engine layers, causing inconsistent behaviour. | Replace raw string fields in `RepositoryCloneRequest` with typed objects: use `&RepositoryRef`, `&BranchName`, and `&WorkspacePath` to enforce validation once at the API boundary. |
| Observability | ⚠️ Warning | No logging, metrics, or tracing at decision points or failure boundaries despite performing operational container Git operations. | Add `tracing::debug!` calls at clone/verification start, all validation failures, and completion. Include container\_id, repository, branch, workspace path, and elapsed time as contextual fields. Add spans for container exec calls. |
| Concurrency And State | ⚠️ Warning | Tests do not exercise concurrency edge cases: no parallel clone calls, multi-threaded runtime scenarios, or cancellation under interleaving despite using async/Tokio. | Add tests for concurrent clones, multi-threaded tokio::test scenarios, nested runtime contexts, and cancellation. Document single-threaded-per-request guarantee in code. |
| Architectural Complexity And Maintainability | ⚠️ Warning | RepositoryCloneRequest uses raw &str fields, triggering duplicate validation. API layer trims branches whilst engine rejects untrimmed input, creating inconsistent semantics across entry paths. | Use typed values in RepositoryCloneRequest fields or move types to shared module to eliminate duplicate validation and establish consistent semantics. |

File: tests/bdd_repository_cloning_e2e_helpers/container.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            return Ok(existing);
+        }
+        if let Some(endpoint) = detect_local_socket() {
+            // SAFETY: Synchronised through `OnceLock::get_or_init`, which
+            // serialises the env mutation across concurrent callers.
+            unsafe {
+                std::env::set_var("DOCKER_HOST", &endpoint);
+            }
+            return Ok(endpoint);
+        }
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/bdd_repository_cloning_e2e_helpers/container.rs:117]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

[the remark this answers is not in the document]
Comment (leynos) [reply tests/bdd_repository_cloning_e2e_helpers/container.rs:117]: @coderabbitai the engineer notes:

> #2 (container.rs DOCKER\_HOST) — Partially addressed. Split into a pure `resolve_docker_host_endpoint` cache (no side effect) and a dedicated `ensure_docker_host` env-mutation gate; the gate now also skips the write when `DOCKER_HOST` is already set in the surrounding environment, and the `SAFETY` comment is strengthened. The residual `unsafe std::env::set_var` is retained because `testcontainers` 0.27 exposes no programmatic Docker host injection — the existing repo helpers (`tests/test_utils.rs::set_env_var`, `tests/bdd_engine_connection_helpers::set_env_var`) either gate on `feature = "internal"` (won't compile into this binary) or store mock-env entries in a `Mutex<HashMap>` rather than mutating the real environment, so neither replaces the testcontainers requirement.

Is further work still required?
